src/Evolution/Systems/ScalarWave/Characteristics.hpp

```diff
@@ -0,0 +1,225 @@
+
+class DataVector;
+
+namespace Tags {
+template <typename Tag>
+struct Normalized;
+}  // namespace Tags
+/// \endcond
+
+namespace ScalarWave {
+namespace Tags {
+struct ConstraintGamma2Compute : ConstraintGamma2, db::ComputeTag {
+  using argument_tags = tmpl::list<Psi>;
+  static auto function(const Scalar<DataVector>& psi) noexcept {
+    return make_with_value<type>(psi, 0.);
+  }
+  using base = ConstraintGamma2;
+};
+}  // namespace Tags
+// @{
+/*!
+ * \ingroup ScalarWave
+ * \brief Compute the characteristic speeds for the scalar wave system.
+ *
+ * Computes the speeds as described in "Optimal constraint projection for
+ * hyperbolic evolution systems" by Holst et. al \cite Holst2004wt
+ * [see text following Eq.(32)]. The characteristic fields' names used here
+ * are similar to the paper:
+ *
+ * \f{align*}
+ * \mathrm{SpECTRE} && \mathrm{Holst} \\
+ * u^{\psi} && Z^1 \\
+ * u^0_{i} && Z^{2}_{i} \\
+ * u^{\pm} && u^{1\pm}
+ * \f}
+ *
+ * The corresponding characteristic speeds \f$v\f$ are given in the text
+ * following Eq.(38) of \cite Holst2004wt :
+ *
+ * \f{align*}
+ * v_{\psi} =& 0 \\
+ * v_{0} =& 0 \\
+ * v_{\pm} =& \pm 1
+ * \f}
+ *
+ * where \f$n_k\f$ is the unit normal to the surface.
+ */
+template <size_t Dim>
+typename Tags::CharacteristicSpeeds<Dim>::type characteristic_speeds(
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unit_normal_one_form) noexcept;
+
+template <size_t Dim>
+void characteristic_speeds(
+    gsl::not_null<typename Tags::CharacteristicSpeeds<Dim>::type*> char_speeds,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unit_normal_one_form) noexcept;
+
+template <size_t Dim>
+struct CharacteristicSpeedsCompute : Tags::CharacteristicSpeeds<Dim>,
+                                     db::ComputeTag {
+  using base = Tags::CharacteristicSpeeds<Dim>;
+  using type = typename base::type;
+  using argument_tags =
+      tmpl::list<::Tags::Normalized<::Tags::UnnormalizedFaceNormal<Dim>>>;
+
+  static typename Tags::CharacteristicSpeeds<Dim>::type function(
+      const tnsr::i<DataVector, Dim, Frame::Inertial>&
+          unit_normal_one_form) noexcept {
+    return characteristic_speeds(unit_normal_one_form);
+  };
+};
+// @}
+
+// @{
+/*!
+ * \ingroup ScalarWave
+ * \brief Computes characteristic fields from evolved fields
+ *
+ * \ref CharacteristicFieldsCompute and
+ * \ref EvolvedFieldsFromCharacteristicFieldsCompute convert between
+ * characteristic and evolved fields for the scalar-wave system.
+ *
+ * \ref CharacteristicFieldsCompute computes
+ * characteristic fields as described in "Optimal constraint projection for
+ * hyperbolic evolution systems" by Holst et. al \cite Holst2004wt .
+ * Their names used here differ from this paper:
+ *
+ * \f{align*}
+ * \mathrm{SpECTRE} && \mathrm{Holst} \\
+ * u^{\psi} && Z^1 \\
+ * u^0_{i} && Z^{2}_{i} \\
+ * u^{\pm} && u^{1\pm}
+ * \f}
+ *
+ * The characteristic fields \f$u\f$ are given in terms of the evolved fields by
```

Comment:
Please use `w_\alpha` when referring to a collection of characteristic variables. This is what we are using in the rest of the evolution code.

Reply:
Actually I believe the GH documentation uses u, presumably, because this is the notation of the GH paper?

Reply:
@fmahebert - yes that is indeed because of the GH paper using `u` and `\hat{u}` for evolved and char fields. I think we should follow the same convention in here as well as in the curved scalar wave system.

Reply:
My issue with `\hat{u}` is that it's standard for denoting the physical solution at grid points, while `w` is rather common in the DG literature as far as I've seen.

Reply:
@nilsdeppe and I discussed this, and would like to use `u_\alpha` for the conserved/evolved variables, `v_{\hat \alpha}` for the associated characteristic variables, `p_\alpha` for primitive variables and `q_{\hat \alpha}` for characteristics of the primitives.

Reply:
Okay, I am now using `v_{\hat \alpha}` for char fields everywhere, and `u` for the evolved fields!